src/repr/src/adt/decimal.rs

```diff
@@ -364,118 +365,133 @@ impl FromStr for Decimal {
+            None => (),
+            Some(b'e') | Some(b'E') => {
+                let mut e_exponent: u8 = 0;
+                let mut e_negative = false;
+
+                // Check for a leading sign for the exponent.
+                if let Some(b'-') = z.peek() {
+                    // Consume the negative sign.
+                    z.next();
+                    e_negative = true;
+                } else if let Some(b'+') = z.peek() {
+                    // Consume the positive sign.
+                    z.next();
+                }
+
+                // Only allow two digits in the exponent, so we don't need to
+                // check for overflow.
+                for _ in 0..2 {
+                    match z.next() {
+                        None => break,
+                        Some(ch @ b'0'..=b'9') => {
+                            e_exponent *= 10;
+                            e_exponent += ch - b'0';
+                        }
+                        Some(_) => bail!("malformed numeric literal: {}", s),
                     }
-                    _ => break,
                 }
-                z.next();
-            }
-        }
+                match z.next() {
+                    None => (),
+                    Some(b'0'..=b'9') => {
+                        bail!("exponent in decimal has more than two digits: {}", s)
+                    }
+                    Some(_) => bail!("malformed numeric literal: {}", s),
+                }
 
-        if z.peek().is_some() {
-            bail!("malformed numeric literal: {}", s);
+                if e_negative {
+                    scale += e_exponent;
+                    if scale > MAX_DECIMAL_PRECISION {
+                        bail!("numeric literal exceeds maximum precision: {}", s);
+                    }
+                } else if scale > e_exponent {
+                    scale -= e_exponent;
+                } else {
+                    e_exponent -= scale;
+                    scale = 0;
+                    let p = 10_u128.checked_pow(e_exponent as u32).ok_or_else(|| {
+                        anyhow!(
+                            "exponent in numeric literal {} overflows i128: 10^{}",
+                            s,
+                            e_exponent
+                        )
+                    })?;
+                    significand = significand
+                        .checked_mul(p)
+                        .ok_or_else(|| anyhow!("numeric literal overflows i128: {}", s))?;
+                }
+            }
+            Some(_) => bail!("malformed numeric literal: {}", s),
         }
 
-        if precision > MAX_DECIMAL_PRECISION {
-            bail!("numeric literal exceeds maximum precision: {}", s);
-        }
+        // Compute the final significand.
+        let mut significand = i128::try_from(significand)
+            .map_err(|_| anyhow!("numeric literal overflows i128: {}", s))?;
         if negative {
             significand *= -1;
         }
```

Comment:
You used to have an check here for `i128::MAX`; I could imagine that tests up above mean that can't happen, but connecting the dots explicitly would help (i.e. the comments above about "not needing to test overflow" aren't as clear about which moment isn't needed, as there are a bunch that might seem needed at first glance but aren't present). There is also `checked_neg()` if you like, which I had not known existed!

Reply:
Correct me if I'm wrong, but I think the reasoning for eliding this specific test is that by constraining the number represented to be at most 38 decimal digits, `i128::MAX` is ruled out along with several other `i128` values just because we aren't right up against 2^127.

Reply:
Oh, yeah, sorry, added that test and later realized it wasn't necessary. I was worried about i128::MAX overflowing when you negate, but that's only an issue when you negate i128::MIN. Added a comment to that effect.